src/linkediting.js

```diff
@@ -55,5 +60,90 @@ export default class LinkEditing extends Plugin {
 
 		// Enable two-step caret movement for `linkHref` attribute.
 		bindTwoStepCaretToAttribute( editor.editing.view, editor.model, this, 'linkHref' );
+
+		// Setup highlight over selected link.
+		this._setupLinkHighlight();
+	}
+
+	/**
+	 * Adds highlight over link which has selection inside, together with two-step caret movement indicates whenever
+	 * user is typing inside the link.
+	 *
+	 * @private
+	 */
+	_setupLinkHighlight() {
+		const editor = this.editor;
+		const model = this.editor.model;
+		const doc = model.document;
+		const highlightDescriptor = {
+			id: 'linkBoundaries',
+			class: 'ck-link_selected',
+			priority: 1
+		};
+
+		// Convert linkBoundaries marker to view highlight.
+		editor.conversion.for( 'editingDowncast' )
+			.add( downcastMarkerToHighlight( {
+				model: 'linkBoundaries',
+				view: highlightDescriptor
+			} ) );
+
+		// Create marker over whole link when selection has "linkHref" attribute.
+		doc.on( 'change', () => {
+			const selection = doc.selection;
+
+			// Create marker over link when selection is inside or remove marker otherwise.
+			if ( selection.hasAttribute( 'linkHref' ) ) {
+				const modelRange = findLinkRange( selection.getFirstPosition(), selection.getAttribute( 'linkHref' ) );
+				const marker = model.markers.get( 'linkBoundaries' );
+
+				if ( !marker || !marker.getRange().isEqual( modelRange ) ) {
+					model.change( writer => {
```

Comment:
Also just today we discussed with PJ, that you shouldn't use `model.change()` in `model.Document#event:change` callback.

I propose switching to `model.enqueueChanges()` and use transparent batch. And see how it works.

Reply:
Or just add this as a post fixer and just use `writer` provided as an argument. In many cases it is quite difficult to decide whether something should be registered as post fixer or a `change` callback.

Reply:
cc @pjasiun

Reply:
Yep, the `change` block should not be used in in the change listener. You can have a postfixer or `enqueueChanges` (even with the same batch)

Reply:
And in this case, I think you should use postfixer.